ldap/servers/slapd/dyncerts.c

```diff
@@ -798,6 +797,786 @@ dyncerts_cert2entry(CERTCertificate *cert)
+        "createTimestamp",
+        "modifyTimestamp",
+        NULL };
+
+    /* Check that dn has the expected format */
+    if (!n->nickname) {
+        PR_snprintf(errmsg, SLAPI_DSE_RETURNTEXT_SIZE,
+                    "Invalid DN for a Dynamic certificate entry."
+                    " DN should be like cn=nickname,%s", DYNCERTS_SUFFIX);
+        return LDAP_NAMING_VIOLATION;
+    }
+    /*
+     * Check to make sure the entry passes the schema check
+     */
+    if (slapi_entry_schema_check(pb, e) != 0) {
+        char *err;
+        slapi_pblock_get(pb, SLAPI_PB_RESULT_TEXT, &err);
+        if (err && err[0]) {
+            PL_strncpyz(errmsg, err, SLAPI_DSE_RETURNTEXT_SIZE);
+        }
+        return LDAP_OBJECT_CLASS_VIOLATION;
+    }
+    /* Check if the attribute values in the entry obey the syntaxes */
+    if (slapi_entry_syntax_check(pb, e, 0) != 0) {
+        char *err;
+        slapi_pblock_get(pb, SLAPI_PB_RESULT_TEXT, &err);
+        if (err && err[0]) {
+            PL_strncpyz(errmsg, err, SLAPI_DSE_RETURNTEXT_SIZE);
+        }
+        return LDAP_INVALID_SYNTAX;
+    }
+    /* Check that only expected attributes are present. */
+    slapi_entry_first_attr(e, &a);
+    for (; a != NULL; slapi_entry_next_attr(e, a, &a)) {
+        Slapi_Value *v = NULL;
+        int i = 0;
+        if (charray_inlist(skipped_attrs, a->a_type)) {
+            continue;
+        }
+        i = slapi_attr_first_value(a, &v);
+        if (i == -1 || slapi_attr_next_value(a, i, &v) != -1) {
+            PR_snprintf(errmsg, SLAPI_DSE_RETURNTEXT_SIZE,
+                        "Attribute %s should be single valued", a->a_type);
+            return LDAP_UNWILLING_TO_PERFORM;
+        }
+        if (!charray_inlist(allowedattrs, a->a_type)) {
+            PR_snprintf(errmsg, SLAPI_DSE_RETURNTEXT_SIZE,
+                        "Unexpected attribute: %s", a->a_type);
+            return LDAP_OBJECT_CLASS_VIOLATION;
+        }
+    }
+    /* Check that nsDynamicCertificateDER attribute is present */
+    if ((needcert || slapi_entry_attr_get_ref(e, DYCATTR_PKEYDER)) &&
+        !slapi_entry_attr_get_ref(e, DYCATTR_CERTDER)) {
+        PR_snprintf(errmsg, SLAPI_DSE_RETURNTEXT_SIZE,
+                    "Attribute %s should be present", DYCATTR_CERTDER);
+        return LDAP_UNWILLING_TO_PERFORM;
+    }
+    return rc;
+}
+
+/* Store the entry in NSS db */
+int
+dyncerts_import_entry(Slapi_Entry *e, char *errmsg, bool verifyOnly)
+{
+    int rc = LDAP_SUCCESS;
+    CertCtx_t ctx = {0};
+
+    dyncert_nickname_from_full_nickname(&ctx.n, slapi_entry_attr_get_ref(e, DYCATTR_NICKNAME));
+    ctx.dercert = slapi_entry_attr_get_secitem(e, DYCATTR_CERTDER);
+    ctx.derpkey = slapi_entry_attr_get_secitem(e, DYCATTR_PKEYDER);
+    ctx.trust = slapi_entry_attr_get_charptr(e, DYCATTR_TRUST);
+    ctx.force = (slapi_entry_attr_get_bool(e, DYCATTR_FORCE) == PR_TRUE);
+    ctx.errmsg = errmsg;
+    ctx.sdn = (Slapi_DN*) slapi_entry_get_sdn_const(e);
```

Comment:
It is a reference in an entry that is later freed. We should duplicate the sdn and free it in dyncerts_cert_ctx_done